File: Residential_Controller.js
/*Method: RequestElevator(RequestedFloor, Direction)
The method 1 must return choosen elevator and make sure to move the elevators in the traitement.

Method
: RequestFloor(Elevator, RequestedFloor)
The method 2 must move the elevators in his traitement*/

//Button Class with THE FloorRequestButton ID
class FloorRequestButton {
    constructor(ID){
        this.ID = ID;
        this.Pressed = false;
    }
}

//Button Class with constructor direction & floor
class Button {
    constructor(direction, floor){
        this.direction = direction;
        this.floor = floor;
        this.light = 'OFF'
    }
}

class Elevator {
    constructor(elevatorID, nbFloor){
        this.elevatorID = elevatorID;
        this.currentFloor = 1;
        this.Direction = 'UP';
        this.RequestList = [];
        this.FloorCallButton = [];
        this.Door = 'CLOSED';
        for (var i = 1; i <= nbFloor; i++){
            this.FloorCallButton.push(i);
        }
    }
}

// Column class with constructor nbEelevator & nbFloor
class Column {
    constructor(ID, nbFloor, nbElevator){
        this.ID = ID;
        this.nbElevator = nbElevator;
        this.nbFloor = nbFloor;
        this.elevatorList = [];
        this.floorList = [];
        this.buttonList = [];
        
        //Add all elevator on the list of Elevator
        for (var i = 1; i <= nbElevator; i++) {
            this.elevatorList.push(new Elevator(i , nbFloor))
        }

         //nomberOfFloors
        for (var i = 1; i <= nbFloor; i++) {
            this.floorList.push(i)
        }
        //callButton
        for (var i = 1; i <= nbFloor; i++) {
            if (i != this.nbFloor - 1 ){
                var callbutton = new Button('DOWN', i)
                this.buttonList.push(callbutton)
                //console.log(this.buttonList);
            }
            if (i != nbFloor ){
                var callbutton = new Button('UP', i)
                this.buttonList.push(callbutton)
                //console.log(this.buttonList);
            }
        }    
    }


    UpdateList(List, currentFloor){
        var List = [];
        List.push(currentFloor);
        List.sort(function(a, b){
            return a-b
        });
    }
    // finding the elevator which is closer to the user position
    nearestElevator (elevatorList, userCurrentFloor){
        var distance = Column.floorList;
        var bestNearestElevator;
        for (let i = 0; i < elevatorList; i++) {
            const elevator = elevatorList[i];
            if (Math.abs(elevator.currentFloor - userCurrentFloor) < distance){                
                bestNearestElevator = elevator;
            }
        }
            return bestNearestElevator;
    }

    RequestElevator(RequestedFloor, direction){
        console.log("The user is at floor " + RequestedFloor + " and is going "  + direction); 
        this.findBestElevator(RequestedFloor, direction);

        //Here's the steps to move the elevator once the user request the elevator
       
            if (Elevator.RequestList > Elevator.currentFloor){
                Elevator.Direction = 'UP';
                while (Elevator.currentFloor < elevator.RequestList[0]){
                    Elevator.currentFloor += 1;
                    console.log("The Elevator # " + elevator.elevatorID + " is now at floor "  + elevator.currentFloor);
                    if (Elevator.currentFloor == Elevator.RequestList){
                        Elevator.Direction = 'IDLE';
                    }
                }
                elevator.Door = 'OPEN';
                console.log('Door is open');
                
            }
            else {
                Elevator.Direction = 'DOWN';
                while (Elevator.currentFloor > Elevator.RequestList){
                    Elevator.currentFloor -= 1;
                    console.log("The elevator No: " + elevator.elevatorID + " is at floor "   + elevator.currentFloor);
                    if (Elevator.currentFloor == 1){
                        Elevator.Direction = 'IDLE';
                    }
                }
                Elevator.Door = 'OPEN';
                console.log('Door is open');
                
            }
            
            Elevator.Door = 'CLOSED';
            console.log('Door is closed');
            Elevator.Direction = 'IDLE';
        
    }
    findBestElevator(RequestedFloor, direction){

            var BestElevator = null ;

       for (let elevator of this.elevatorList) {
            console.log("The Elevator #" + elevator.elevatorID + " is now at floor "  + RequestedFloor + ' floor and its direction is ' + direction);
            if (elevator.currentFloor == RequestedFloor && elevator.Direction == direction){
                if (elevator.Door == 'OPEN'){
                    this.UpdateList(elevator.RequestList, RequestedFloor);
                    console.log("the best elevator No: " + elevator.elevatorID + ' is there!');
                    return elevator;
                }
            }
            else if(elevator.currentFloor > RequestedFloor && elevator.Direction ==='IDLE'){           
                BestElevator = elevator;
            }
            else if(direction === 'UP' && elevator.currentFloor < RequestedFloor){
                BestElevator = elevator;
            }
            else if(direction === 'DOWN' && elevator.currentFloor > RequestedFloor){
                BestElevator = elevator;
            }
            else if(direction === 'DOWN' && elevator.Direction === 'DOWN'){
                BestElevator = elevator;
            }
            else if(direction === 'UP' && elevator.Direction === 'UP'){
                BestElevator = elevator;
            }
            else{
                BestElevator = this.nearestElevator( BestElevator, direction);
                this.UpdateList(this.RequestList, RequestedFloor);
                console.log("The elevator No : " + elevator.elevatorID + ' is comming!');
                return BestElevator;
            }


        }
        
    }
    RequestFloor(elevator, RequestedFloor){ 
        this.UpdateList(elevator, RequestedFloor);

        //Here's the steps to move the elevator once the user is in the elevator
       
            if (Elevator.RequestList > Elevator.currentFloor){
                Elevator.Direction = 'UP';
                while (elevator.currentFloor < Elevator.RequestList[0]){
                    Elevator.currentFloor += 1;
                    console.log("The Elevator # " + elevator.elevatorID + " is now at floor "  + elevator.currentFloor);
                    if (Elevator.currentFloor == Elevator.RequestList){
                        Elevator.Direction = 'IDLE';
                    }
                }
                Elevator.Door = 'OPEN';
                console.log('Door is opened');
                
            }
            else {
                Elevator.Direction = 'DOWN';
                while (Elevator.currentFloor > Elevator.RequestList){
                    Elevator.currentFloor -= 1;
                    console.log("The elevator No: " + elevator.elevatorID + " is at floor "   + elevator.currentFloor);
                    if (elevator.currentFloor == 1){
                        elevator.Direction = 'IDLE';
                    }
                }
                Elevator.Door = 'OPEN';
                console.log('Door is open');
                
            }
            
            Elevator.Door = 'CLOSED';
            console.log('Door is closed');
            Elevator.Direction = 'IDLE';
        
    }   
}


// --- /Classes ----   
// --- Scenarios ---    
function Scenario1(){
    console.log('******************* ******************* *******************');
    console.log('*******************      Scenario 2     *******************');
    console.log('******************* ******************* *******************');
    column1 = new Column(1, 10, 2);

    column1.elevatorList[0].currentFloor = 1;
    column1.elevatorList[0].Direction = 'IDLE';
    column1.elevatorList[1].currentFloor = 7;
    column1.elevatorList[1].Direction = 'IDLE';
    console.log('******************* USER-1 goes from floor 1 to floor 6  *******************');
    RequestedFloor = 2;
    Direction = 'DOWN';
    Destination = 6;
    elevatorList = [1, 2];
    column1.RequestElevator(RequestedFloor, Direction);
    column1.nearestElevator(elevatorList, 1);
    column1.RequestFloor(2, 7);
    console.log('******************* USER-3 goes from floor 9 to floor 2  *******************');
    RequestedFloor = 9;
    Direction = 'DOWN';
    Destination = 2;
    elevatorList = [1, 2];
    column1.RequestElevator(RequestedFloor, Direction);
    column1.nearestElevator(elevatorList, 9);
    column1.RequestFloor(2, 2);
}

/*column = new Column(1, 10, 2);
best = column.findBestElevator(3, 'DOWN')
console.log(best)
best2 = column.findBestElevator(6, 'UP')
console.log(best2)

elevator = column.RequestFloor(1, 7)
console.log(elevator)
console.log(column)*/
//  --- /Scenarios---

Scenario1();

File: Residential_Controller.py
"""Method: RequestElevator(RequestedFloor, Direction)
The method 1 must return choosen elevator and make sure to move the elevators in the traitement.

Method
: RequestFloor(Elevator, RequestedFloor)
The method 2 must move the elevators in his traitement"""

#Button Class with THE FloorRequestButton ID
class FloorRequestButton():
    def __init__(self, ID):
        self.ID = ID
        self.Pressed = False

#Button Class with constructor direction & floor
class Button() :
    def __init__ (self, direction, floor):
        self.direction = direction
        self.floor = floor
        self.light = 'OFF'

#Elevator Class with the elevator & floor
class Elevator():
    def __init__(self, ID, nbFloor):
        self.ID = ID
        self.currentFloor = 1
        self.direction = 'IDLE'
        self.requestList = [] 
        self.floorCallButton = []
        self.door = 'CLOSED'
        for i in range (nbFloor):
            self.floorCallButton.append(i)


# Column class with constructor nbEelevator & nbFloor
class Column(): 
    def __init__ (self, ID, nbFloor, nbElevator):
        self.ID = ID
        self.nbElevator = nbElevator
        self.nbFloor = nbFloor
        self.elevatorList = []
        self.floorList = []
        self.buttonList = []

       
        #Add all elevator on the list of Elevator
        for i in range (nbElevator):
            elevator = Elevator(i + 1, nbFloor)
            self.elevatorList.append(elevator)
            #print(self.elevatorList)

        #nomberOfFloors
        for i in range (nbFloor):
            self.floorList.append(i)

        #callButton
        for i in range (nbFloor):
            if i != self.nbFloor - 1 :
                callbutton = Button('DOWN', i)
                self.buttonList.append(callbutton)
               

            if i != nbFloor :
                callbutton = Button('UP', i)
                self.buttonList.append(callbutton)

    def UpdateList(self, elList, currentFloor):
        elList = []
        elList.append(currentFloor)
        elList.sort()

    #finding the elevator which is closer to the user current floor
    def nearestElevator (self, elevatorList, userCurrentFloor):
        distance = len(self.floorList)
        bestNearestElevator = elevatorList[0]
        for elevator in elevatorList:
            if abs(elevator.currentFloor - userCurrentFloor) < distance:
                bestNearestElevator = elevator
        return bestNearestElevator

    def RequestElevator(self, requestedFloor, direction):
        self.findBestElevator(requestedFloor, direction)
        return requestedFloor

    def findBestElevator(self, requestedFloor, direction):
        print("The user is at floor ", requestedFloor , " and is going ", direction)
        BestElevator = []
        for elevator in self.elevatorList:
            print("The Elevator #", elevator.ID , " is now at floor ", elevator.currentFloor , " and its direction " , elevator.direction)
            if(elevator.currentFloor == requestedFloor and elevator.direction == direction):
                if(elevator.door == 'OPEN'):
                    self.UpdateList(elevator.requestList, requestedFloor)
                    print("The best elevator No" , elevator.ID , " is there!")
                    self.moveElevator(elevator)
                    return elevator
            elif (elevator.currentFloor == direction):
                if (elevator.currentFloor > requestedFloor and elevator.direction == 'DOWN'):
                    BestElevator.append(elevator)
                elif (elevator.currentFloor < requestedFloor and elevator.direction == 'UP'):
                    BestElevator.append(elevator)
                else:
                    BestElevator.append(elevator)
            elif (elevator.direction == 'IDLE'):
                if (elevator.currentFloor > requestedFloor and elevator.direction == 'IDLE'):
                    BestElevator.append(elevator)        
                elif (elevator.currentFloor < requestedFloor and elevator.direction == 'IDLE'):
                    BestElevator.append(elevator)
                else:
                    BestElevator.append(elevator)   
            else:
                bestElevator = self.nearestElevator(BestElevator , requestedFloor)
                self.UpdateList(elevator.requestList, requestedFloor)
                print("The elevator No" , elevator.ID , " is comming")
                self.moveElevator(bestElevator)
                return bestElevator
    
    def moveElevator(self, elevator):
        while (len(elevator.requestList) > 0):
            if(elevator.requestList[0] > elevator.currentFloor):
                elevator.direction = 'UP'
                while (elevator.currentFloor < elevator.requestList[0]):
                    elevator.currentFloor += 1
                    print("The elevator #" , elevator.ID , " is now at floor ", elevator.currentFloor)

                    if (elevator.currentFloor == len(elevator.requestList[0])):
                        elevator.direction = 'IDLE'
            else:
                elevator.direction = 'DOWN'
                while (elevator.currentFloor > elevator.requestList[0]):
                    elevator.currentFloor -= 1
                    print("The elevator #" , elevator.ID , " is at floor " , elevator.currentFloor)
                    if (elevator.direction == 1):
                        elevator.direction = 'IDLE' 
                elevator.door = 'OPEN'
                print("Door opened")
            
            elevator.door = 'CLOSED'
            print("Door is closed")   
    
    def RequestFloor(self, elevator, requestedFloor):
            self.UpdateList(elevator.requestList, requestedFloor)
            self.moveElevator(elevator)
            return elevator



# --- Scenarios ---
def Scenario1 ():
    print('******************* ******************* *******************')
    print('*******************      Scenario 1     *******************')
    print('******************* ******************* *******************')
    column = Column(1, 10, 2)

    column.elevatorList[0].CurrentFloor = 2
    column.elevatorList[0].Direction = 'IDLE'
    column.elevatorList[1].CurrentFloor = 6
    column.elevatorList[1].Direction = 'IDLE'
    print('******************* USER-1 goes from floor 3 to floor 7  *******************')
    RequestedFloor = 3
    direction = 'UP'
    destination = 7
    elevator = column.RequestElevator(RequestedFloor, direction)
    column.RequestFloor(elevator, destination)
    
# --- /Scenarios---
Scenario1() 


